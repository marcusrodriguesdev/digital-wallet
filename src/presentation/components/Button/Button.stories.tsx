import React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import '@react-theming/storybook-addon/register'
import { ThemeProvider } from 'styled-components'
import { DarkTheme } from '@/presentation/theme/dark.theme'
import { Button } from './Button'

storiesOf('presentation/components/Button', module)
  .addDecorator((storyFn) => (
    <div style={{ backgroundColor: '#262626', padding: '20px' }} className="auth__container__content">
      {storyFn()}
    </div>
  ))
  .add('default', () => (
    <ThemeProvider theme={DarkTheme}>
      <Button scale="medium" label="Entrar" variant="default" onClick={action('onClick')} />
    </ThemeProvider>
  ))
