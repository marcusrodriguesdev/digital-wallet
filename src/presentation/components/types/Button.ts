export type ButtonVariantTypes = 'outline' | 'default'

export type ScaleTypes = 'small' | 'medium' | 'large'
