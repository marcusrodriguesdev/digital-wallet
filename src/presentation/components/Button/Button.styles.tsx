import styled from 'styled-components'
import { ButtonVariantTypes, ScaleTypes } from '../types/Button'

interface ContainerProps {
  variant: ButtonVariantTypes
  scale: ScaleTypes
  disabled: boolean
}

export const Container = styled.button<ContainerProps>`
  align-items: center;
  background-color: ${({ theme, variant }) =>
    variant === 'outline' ? 'transparent' : theme?.colors?.highlight ?? '#7f5af0'};
  border: ${({ theme }) => theme?.colors?.highlight ?? '#7f5af0'} solid 2.5px;
  border-radius: 2px;
  cursor: pointer;
  display: flex;
  height: ${({ scale: sized }) => (sized === 'small' ? 40 : sized === 'medium' ? 45 : 50)}px;
  justify-content: center;
  opacity: ${({ disabled }) => (disabled ? '0.4' : '1')};
  width: ${({ scale: sized }) => (sized === 'small' ? 240 : sized === 'medium' ? 270 : 300)}px;

  &:hover {
    opacity: 0.8;
  }
`

interface TextProps {
  scale: ScaleTypes
}

export const Text = styled.span<TextProps>`
  ${({ theme }) => theme?.fonts?.bigInfo ?? {}}
  background-color: transparent;
  color: ${({ theme }) => theme?.colors?.main ?? '#fffffe'};
`
