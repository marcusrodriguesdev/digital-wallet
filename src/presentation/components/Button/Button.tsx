import React from 'react'

import { ButtonVariantTypes, ScaleTypes } from '../../types/Button'
import { Container, Text } from './Button.styles'

export interface ButtonProps {
  label: string
  variant: ButtonVariantTypes
  scale: ScaleTypes
  disabled?: boolean
  onClick: () => void
}

/**
 * A stylized button for the application
 *
 * @param label     The Buttons's display text
 * @param variant   Control Button shape style
 * @param scale     Control Button size
 * @param disabled  Disable button click and change it's opacity
 * @param onClick   Callback on item press
 */

const Button: React.FC<ButtonProps> = (props: ButtonProps) => {
  const { label, variant, disabled, scale, onClick } = props

  const handlePress = (): void => {
    if (!disabled) onClick()
  }

  return (
    <Container
      type="button"
      data-testid="Button/Button"
      variant={variant}
      disabled={disabled}
      onClick={handlePress}
      scale={scale}>
      <Text scale={scale}>{label}</Text>
    </Container>
  )
}

export { Button }
