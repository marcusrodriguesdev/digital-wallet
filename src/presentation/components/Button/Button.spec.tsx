import React from 'react'

import { fireEvent, render } from '@testing-library/react'
import * as Faker from 'faker'

import { Button, ButtonProps } from './Button'
import '@testing-library/jest-dom/extend-expect'

const makeSUT = (props?: Partial<ButtonProps>): any => {
  const label = props?.label || Faker.random.words(2)
  const variant = props?.variant || 'default'
  const scale = props?.scale || 'medium'
  const disabled = props?.disabled || false
  const onClickSpy = jest.fn()

  const sut = <Button label={label} variant={variant} scale={scale} disabled={disabled} onClick={onClickSpy} />

  return { sut, label, variant, scale, disabled, onClickSpy }
}

describe('presentation/components/Button', () => {
  it('should match snapshot', () => {
    const { sut } = makeSUT({
      label: 'Entrar',
      variant: 'default',
      scale: 'medium'
    })
    expect(sut).toMatchSnapshot()
  })

  it('should call onClickSpy when button is clicked', () => {
    const { sut, onClickSpy } = makeSUT()

    const { getByTestId } = render(sut)
    const button = getByTestId('Button/Button')

    fireEvent.click(button)

    expect(onClickSpy).toBeCalledTimes(1)
  })

  it('should not call onClickSpy when button is clicked and disabled is true', () => {
    const { sut, onClickSpy } = makeSUT({ disabled: true })

    const { getByTestId } = render(sut)
    const button = getByTestId('Button/Button')

    fireEvent.click(button)

    expect(onClickSpy).toBeCalledTimes(0)
  })

  it('should render with correct label', () => {
    const { sut, label } = makeSUT()

    const { getByTestId } = render(sut)
    const button = getByTestId('Button/Button')

    expect(button.textContent).toBe(label)
  })

  it('should render with correct scale', () => {
    const { sut, scale } = makeSUT({
      scale: 'small'
    })

    const { getByTestId } = render(sut)
    const button = getByTestId('Button/Button')

    expect(button).toHaveAttribute('scale', scale)
  })
})
