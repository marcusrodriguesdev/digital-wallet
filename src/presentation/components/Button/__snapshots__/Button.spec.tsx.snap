// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`presentation/components/Button should match snapshot 1`] = `
<Button
  disabled={false}
  label="Entrar"
  onClick={[MockFunction]}
  scale="medium"
  variant="default"
/>
`;
